### ISSUE TITLE
Passing Jaas config parameter in Spark-shell

### ISSUE THREAD

**Issue by Prophet32j** [lucidworks/spark-solr#168]
I'm trying to use the shaded 2.0.4 jar with spark-shell, but I can't get the class to pick up the additional CLI argument. Is there a trick to getting Kerberos and spark-solr to work with spark-shell?

```
spark-shell \
--jars /san-data/itadausr/lib/spark-solr-2.0.4-shaded.jar \
--conf "spark.executor.extraJavaOptions=-DsolrJaasAuthConfig=/san-data/itadausr/solr/config/jaas-client-test.conf" \
--conf "spark.driver.extraJavaOptions=-DsolrJaasAuthConfig=/san-data/itadausr/solr/config/jaas-client-test.conf" \
--files "/san-data/itadausr/solr/config/jaas-client-test.conf,/opt/cst/spnego/my.keytab"
```
```
val zkhost = "my.zkhost:8893"
val collection = "my-collection"

import com.lucidworks.spark.rdd.SolrRDD
val solrRDD = new SolrRDD(zkhost, collection ,sc)
```

throws a 401 error, not authenticated.

**Comment by kiranchitturi** [comment]
Your usage is wrong I think. The code looks for this system parameter https://github.com/lucidworks/spark-solr/blob/master/src/main/scala/com/lucidworks/spark/util/SolrSupport.scala#L100

which turns out to be `java.security.auth.login.config` (See https://github.com/apache/lucene-solr/blob/master/solr/solrj/src/java/org/apache/solr/client/solrj/impl/Krb5HttpClientBuilder.java#L52)

The example listed here https://github.com/LucidWorks/spark-solr#kerberos only works if you build your application through `com.lucidworks.spark.SparkApp` class. It is a bit misleading. I will update the spark-shell example there

**Comment by Prophet32j** [comment]
what does that mean `build your application through com.lucidworks.spark.SparkApp`? Is there an example of this that I can reference?

**Comment by kiranchitturi** [comment]
There are few examples, take a look at https://github.com/lucidworks/spark-solr/tree/master/src/main/java/com/lucidworks/spark/example

**Comment by Prophet32j** [comment]
Ok. I see how that example is supposed to work. We extend SparkApp.RDDProcessor, but now I'm getting a different error that it can't find a main method.
```scala
class AssetRetirer extends SparkApp.RDDProcessor {
  
  def getName: String = "asset-retirer-example"
  
  def getOptions: Array[Option] = Array()
  
  def run(conf: SparkConf, cli: CommandLine): Int = {
    
    val sc = SparkContext.getOrCreate(conf)
    val zkhost = "my-server:8893"
    val collection = "my-assets"
    val solrQuery = new SolrQuery("body_t:ITADA")
    val solrRDD = new SolrRDD(zkhost, collection ,sc)
    val results = solrRDD.query(solrQuery)
    val count = results.count
    val stringRDD = results.map(_.toString)
    stringRDD.saveAsTextFile("/data/applications/itadausr/staging/SolrTest")
    
    count.toInt
  }
}
```
```
Exception in thread "main" java.lang.NoSuchMethodException: sf.itmp.itada.util.spark.AssetRetirer.main([Ljava.lang.String;)
	at java.lang.Class.getMethod(Class.java:1678)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:716)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:181)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:206)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

**Comment by Prophet32j** [comment]
ok I think I understand what it means to build through. You mean run through the SparkApp, so that SparkApp is our entry point that will call our implementing RDDProcessor class. So now I reconfigured the spark-submit to this:
```
spark-submit \
--class com.lucidworks.spark.SparkApp \
--master yarn-client \
--jars itada-spark-utils-ufen.jar \
spark-solr-2.0.4-shaded.jar \
sf.itmp.itada.util.spark.AssetRetirer \
-solrJaasAuthConfig=/san-data/itadausr/solr/config/jaas-client-test.conf
```
and I get the following error now
```
Exception in thread "main" java.lang.AbstractMethodError: sf.itmp.itada.util.spark.AssetRetirer.getOptions()[Lshaded/apache/commons/cli/Option;
	at com.lucidworks.spark.SparkApp.main(SparkApp.java:142)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:731)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:181)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:206)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

**Comment by Prophet32j** [comment]
We found the errors here with this project shading and relocating jars and our cluster having the normal packages already available in the CDH bundle. So now we are working to shade and relocate as well to match the classes of this repo.

**Comment by Prophet32j** [comment]
I think we have all of the shading figured out. We are back to getting a 401 error which is a good thing. Now I think we may have our endpoint wrong.

The endpoint, is that supposed to be the ZooKeeper or is that the REST API endpoint? With ZK, we have like 7 different servers.

**Comment by kiranchitturi** [comment]
+1 on figuring out the shading

What endpoint are you talking about?

**Comment by Prophet32j** [comment]
```
spark-submit --master yarn-server \
  --class com.lucidworks.spark.SparkApp \
  $SPARK_SOLR_PROJECT/target/spark-solr-${VERSION}-shaded.jar \
  hdfs-to-solr -zkHost $ZK -collection spark-hdfs \
  -hdfsPath /user/spark/testdata/syn_sample_50k \
  -solrJaasAuthConfig=/path/to/jaas-client.conf
```
`-zkHost $ZK` by parameter tells me that it wants the zookeeper endpoint, not a REST API endpoint. I thought SolrJ uses the REST API to call into Solr, since it's creating an HTTP GET request.

in the examples everything is pointing to localhost, but that's not how things work in production. We have a lot of servers operating in our cluster.

**Comment by Prophet32j** [comment]
We also found #83 that may have something to do with Kerberos not working correctly? I bumped our version of spark-solr to 2.1.0 and I got this error, which means something isn't compatible:
```
Exception in thread "main" java.lang.UnsupportedClassVersionError: com/lucidworks/spark/SparkApp : Unsupported major.minor version 52.0
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:803)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:449)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:71)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:278)
	at org.apache.spark.util.Utils$.classForName(Utils.scala:175)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:689)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:181)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:206)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

**Comment by kiranchitturi** [comment]
-zkHost $ZK by parameter tells me that it wants the zookeeper endpoint, not a REST API endpoint. I thought SolrJ uses the REST API to call into Solr, since it's creating an HTTP GET request.

yes, it should be ZK connect string of your SolrCloud instance

**Comment by kiranchitturi** [comment]
yeah, you might have to backport and build your own spark-solr version for that fix

**Comment by Prophet32j** [comment]
we copied the fix code into our project and shaded the fix into 2.0.4 without having to build our own version. That worked fine, however we are still back in the same problem of getting back 401 errors, but now it's 401 errors after negotiating with the zookeeper host.

I think we are going to have to try another solution. We wanted to use this repo, but we are tied to Cloudera releases for our cluster, and this project has different versions than what we are on right now. Not sure if that's the true issue or it's just the fact that Kerberos is a nightmare to work with in the Hadoop cluster.

Closing.
